### PR TITLE
Allow exporting data encoded value as parameter for ply-ctl

### PR DESCRIPTION
**Pull request by SeungheonOh**

### CODE REVIEW

**Comment by TotallyNotChase** [comment]
I'm not well versed enough in CTL to appropriately review those parts. But the Haskell part looks fine.

I do wonder however, given the size of the additions (compared to the existing subprojects), perhaps we should make a new repo for `ply-ctl`?

**Comment by SeungheonOh** [comment]
I made [ply-ctl](https://github.com/mlabs-haskell/ply-ctl).

**Review by TotallyNotChase — COMMENTED** [review]
Some minor comments.

Can you also add some documentation regarding genPurs to the README>

File: .gitignore [hunk 1]
```diff
@@ -10,6 +10,29 @@ node_modules/
 dist/
 dist-newstyle/
 
+# Purescript artifact files
```
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply .gitignore:13]: yep.

**Comment by SeungheonOh** [comment]
New `PlyTypeName` is quite good. It encode `Symbol` from `Tagged`. So we get this:
```json
["Ply.Core.Types:AsData#GHC.Num.Integer:Integer",
"Ply.Core.Types:AsData#PlutusLedgerApi.V1.Value:CurrencySymbol",
"Ply.Core.Types:AsData#Data.Tagged:Tagged#GHC.TypeLits:\"BatchToken\"#Plutarch.Extra.AssetClass:AssetClass",
"Ply.Core.Types:AsData#Data.Tagged:Tagged#GHC.TypeLits:\"ActionToken\"#Plutarch.Extra.AssetClass:AssetClass",
"Ply.Core.Types:AsData#Data.Tagged:Tagged#GHC.TypeLits:\"Underlying\"#Plutarch.Extra.ExtendedAssetClass:ExtendedAssetClass",
"Ply.Core.Types:AsData#PlutusLedgerApi.V1.Scripts:ScriptHash",
"Ply.Core.Types:AsData#Data.Tagged:Tagged#GHC.TypeLits:\"MarketParams\"#Plutarch.Extra.AssetClass:AssetClass",
"Ply.Core.Types:AsData#Data.Tagged:Tagged#GHC.TypeLits:\"MarketState\"#Plutarch.Extra.AssetClass:AssetClass"
]
```

**Comment by SeungheonOh** [comment]
I'll add proper documentation next week once I finish migrating `liqwid-offchain` to use ply-ctl

File: ply-core/src/Ply/Core/Types.hs [hunk 2]
```diff
@@ -38,6 +39,8 @@ import UntypedPlutusCore (DeBruijn, DefaultFun, DefaultUni, Program)
 import Ply.Core.Serialize.Script (serializeScriptCbor)
 import Ply.Core.Typename (Typename)
 
+data AsData a = AsData a
```
Comment (MangoIV) [inline ply-core/src/Ply/Core/Types.hs:42]: Is there a particular reason why this is not a `newtype`?

**Review by MangoIV — COMMENTED** [review]
added a little bike-shed because I was reading this anyway because I think I need it, thank you!

File: ply-core/src/Ply/Core/Types.hs (hunk 2 above)
Comment (SeungheonOh) [reply ply-core/src/Ply/Core/Types.hs:42]: Nice catch. I was planning on different thing.

File: genPurs/genPurs.cabal [hunk 3]
```diff
@@ -0,0 +1,92 @@
+cabal-version: 3.0
+name:          ply-ctl
+version:       0.4.0
+author:        Seungheon <Seungheon@mlabs.city>
+license:       MIT
+
+-- Common sections
+
+common common-configs
+  default-language: Haskell2010
```
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply genPurs/genPurs.cabal:10]: Personally, I like to manually enable extensions, just to be explicit. That being said, GHC2021 should be good too

File: example/reader-app/Main.hs [hunk 4]
```diff
@@ -2,12 +2,25 @@ module Main (main) where
...
 data MintingPolicy = MintingPolicy (Program DeBruijn DefaultUni DefaultFun ())
 
+instance Show MintingPolicy where
+  show =
+    Text.unpack
+      . Text.decodeUtf8
+      . Base16.encode
+      . SBS.fromShort
+      . serialiseUPLC
+      . (\(MintingPolicy x) -> x)
```
[the remark this answers is not in the document]
Comment (SeungheonOh) [reply example/reader-app/Main.hs:22]: It's for testing only, I don't think it's important

File: genPurs/genPurs.cabal (hunk 3 above)
[the remark this answers is not in the document]
Comment (MangoIV) [reply genPurs/genPurs.cabal:10]: Idk; it seems arbitrary to me to use `Haskell2010` to enable a bunch of extensions and then not use `GHC2021` for extensions you'll enable anyway, but you do you.

File: ply-core/src/Ply/Core/Types.hs (hunk 2 above)
Comment (MangoIV) [reply ply-core/src/Ply/Core/Types.hs:42]: this should be caught by the linter, this should be part of pre-commit-hooks. I will probably open a PR with some nix fixes if I find time...

**Comment by SeungheonOh** [comment]
Sorry for delay, I was pretty sick last week. This PR is ready to be merged. 

github.com/mlabs-haskell/ply-ctl is pretty much ready for production. I've replaced old script exporting system with ply-ctl for agora-offchain, and works quite well.

**Comment by SeungheonOh** [comment]
Changelog and documentations' been updated as well

**Review by TotallyNotChase — APPROVED** [review]

**Pull request merged**